Fixes #81 - Aging Not Applying

doodlebunnyhops opened the pull request:
Fixes #81 
Updated sections looking for `produce.keg != null` where it should be `produce.kegType != null &&  options.aging != "None"` to see if aging is applied.

doodlebunnyhops committed:
check for kegType using incorrect produce type

Doksperiments commented:
Hey i was having the same issue that was reported and ran into your pull request.

I'm pretty sure the solutions you implemented won't work.
On line 446 you changed the check to jarType but forgot to change the content of the statement. As you have it now it will return a string of the jarType instead of calculating the actual income.

A fixed version could look like this: 
`netIncome += itemsMade * (crop.produce.jarType != null ? options.skills.arti ? (crop.produce.price * 2 + 50) * 1.4 : crop.produce.price * 2 + 50 : 0);`

In the solution for the kegs, the condition works only if the aging is null, otherwise the artisan bonus will still be counted twice (which was the original error causing incorrect values) and in the case of fixed price products (like beer), the price will be entirely incorrect (since you aren't using the produce.keg value anymore).

One solution I tried to implement is to remove the artisan bonus calculations from getKegModifier entirely, since the function isn't ever called alone, and the artisan bonus is included in the caskModifier.
 
The function getKegModifier could be simplified to:
`return crop.produce.kegType == "Wine" ? 3 : 2.25;`

And the code for calculating the income and kegPrice would look like this:
`netIncome += itemsMade * (crop.produce.keg != null ? crop.produce.keg * caskModifier : Math.floor(crop.produce.price * kegModifier) * caskModifier);`
`var kegPrice = d.produce.keg != null ? d.produce.keg * caskModifier : Math.floor(d.produce.price * kegModifier) * caskModifier;
`

I tried to double check the new values with the wiki and ingame values and they seem to match, but there could be some edge cases i missed

doodlebunnyhops committed:
imporperly returned jarType, update keg/cask forumla

doodlebunnyhops commented:
Good catch @Doksperiments - I updated jar to return 0 on type null.

On your other point for keg/cask. I reevaluated the original formula here and I had an oversight on the use of `crop.produce.keg` - being available to items that have a different base value other `base*3`. 

I updated the formula to first not apply artisan buff till much later and have `getKegModifier` accounting for the edge case of Coffee and Tea, lastly `getCaskModifier` simplified to just aging.

Additionally I repurposed `var kegPrice` that was calculated in the drawing function to be set as `profitData.kegPrice` to reduce function calling on line 456.

Doksperiments reviewed (commented):
Figured a code review would be better to highlight specific parts of the code instead of writing a single comment

Doksperiments commented on js/main.js:
```diff
@@ -442,16 +441,19 @@ function profit(crop) {
                 else if (options.produce == 2) {
-                    netIncome += itemsMade * (crop.produce.kegType != null && options.aging != "None" ? crop.produce.price * kegModifier * caskModifier : crop.produce.price * kegModifier);
+					kegPrice = crop.produce.keg != null ? crop.produce.keg * kegModifier : crop.produce.kegType != null && options.aging != 0 ? crop.produce.price * kegModifier * caskModifier : crop.produce.price * kegModifier;
+                    netIncome += options.skills.arti ? itemsMade * (kegPrice * 1.4) : itemsMde * kegPrice;
```
To add to this, currently the calculation for coffee is wrong, because it calculates the `netIncome` by multiplying the number of beans produced by the price of a single cup, ignoring the fact that 5 beans are necessary for a single coffee to be made, but it would be a separate issue that can be fixed once the general pricing logic is correct.

doodlebunnyhops committed:
use kegBasePrice to simplifiy keg/cask conditionals

doodlebunnyhops commented:
@Doksperiments - would appreciate your input on changes made. 
@Thorinair you as well :D

- crops.js: for produce that can be used in a keg/cask a boolean of `ages` has been added. 
- main.js
  - `getKegModifier` removed
  - `getKegBasePrice` added
  - `kegPrice` simplified to `kegPrice = crop.produce.ages ? kegBasePrice * caskModifier : kegBasePrice;`

Doksperiments commented on js/main.js:
```diff
@@ -203,32 +203,34 @@ function levelRatio(fertilizer, level, isWildseed) {
-    }
+function getCaskModifier(crop) {
+	var modifier = 1
+	if (options.aging > 0 && crop.produce.ages){
```
Just nitpicking, but you could do without the `options.aging > 0` since the default case can handle that (and the options set in the html file only allow numbers from 0 to 3)

Doksperiments commented on js/main.js:
```diff
@@ -1133,7 +1138,7 @@ function renderGraph() {
 					if (d.produce.jarType) {
 						tooltipTr.append("td").attr("class", "tooltipTdLeftSpace").text("Value (" + d.produce.jarType + "):");
-						tooltipTr.append("td").attr("class", "tooltipTdRight").text(d.produce.price * 2 + 50)
+						tooltipTr.append("td").attr("class", "tooltipTdRight").text(options.skills.arti ? Math.round((d.produce.price * 2 + 50) * 1.4) : d.produce.price * 2 + 50)
```
I'd say to keep this in the other pull request just to avoid confusion and not mix two distinct issues into one.
On that note, maybe it would make sense to change the name of this pull request to encompass the wider 'keg calculations' problem

Doksperiments commented on js/main.js:
```diff
@@ -203,32 +203,34 @@ function levelRatio(fertilizer, level, isWildseed) {
+				break;
+			case 3: modifier = 2;
+				break;
+			default: modifier = 1;
```
Again, nitpicking, but you could just break on the default case since modifier is initialized at 1.
I would advise against removing the case entirely since it can still catch unexpected values and just for future-proofing sake.

Doksperiments reviewed (commented):
I tried a few of cases which were confirmed wrong before and the values seem to be all correct, but as always with cherry-picked tests there could be some cases I missed (which hopefully will be reported if they get found).
It would probably be useful to have a test suite to test against, but I feel no one would want to actually write it :D  

Other than that it seems to me that it's working as intended now

doodlebunnyhops committed:
getCaskModifier responsible for age check, removed jar edits

rangigo commented:
Hey I just realised this PR fixed the problem showing incorrect Keg price with `caskModifier`. Do you know when this can be merged? Does only @Thorinair have the merge access for this repo?

Thorinair commented:
If others believe that it is ready for a merge, I could merge it this week!

doodlebunnyhops commented:
> If others believe that it is ready for a merge, I could merge it this week!

I'm ready for it ☺️

a-johanes commented:
the codes looks great! not sure if we should consider the number of days into calculation for aging?

a-johanes commented:
> If others believe that it is ready for a merge, I could merge it this week!

other than the above, I think the PR is ready for merge @Thorinair !

Thorinair commented:
Hey, would you please be able to have a look at this again? I have merged a bunch of pull requests, including @doodlebunnyhops prediction model, so there are now come conflicts. Notable things I have merged are the prediction model and support for Mill produce type. Conflicts in `crops.js` are easy to merge, you just need to keep both and fix the commas, but the ones in `main.js` are are more different because at some point, some parts of the calculation were changed.

doodlebunnyhops commented:
> Hey, would you please be able to have a look at this again? I have merged a bunch of pull requests, including @doodlebunnyhops prediction model, so there are now come conflicts. Notable things I have merged are the prediction model and support for Mill produce type. Conflicts in `crops.js` are easy to merge, you just need to keep both and fix the commas, but the ones in `main.js` are are more different because at some point, some parts of the calculation were changed.

Ah sorry been a bit I'll check merge issues!

doodlebunnyhops committed:
Merge upstream/dev into dev-aging

doodlebunnyhops commented:
Updated to resolve merge conflicts, would love more testing from others!
@Thorinair

Thorinair commented:
Oh dear, you did lots of work here! Thank you! I will have a look tomorrow and merge it up! :D

Thorinair commented:
Pushed to live now!
